### PR TITLE
CNI Support

### PR DESCRIPTION
**Pull request by wwwtyro**
@chuckbutler @mbruzek @Cynerva @marcoceppi 

Here's a [brief mention](http://kubernetes.io/docs/admin/network-plugins/#cni) of the CNI plugin in the Kubernetes documentation. 

Here's some [instructions](http://docs.projectcalico.org/v1.6/getting-started/kubernetes/installation/#manual-installation) for setting up Calico CNI with Kubernetes. 

The basic workflow for Calico CNI setup is as follows:

1. Set up a systemd service for the **calicoctl** utility and run it on both master and every worker. This utility requires Docker and a connection to etcd.
2. On each worker, install the Calico CNI plugins (**calico** and **calico-ipam**) and a configuration file. The configuration file requires etcd connection information and the location of the kubeconfig file.
3. On each worker, install the CNI lo plugin (**loopback**).
4. Deploy the Calico network policy controller by downloading the [manifest](http://docs.projectcalico.org/v1.6/getting-started/kubernetes/installation/policy-controller.yaml), setting the ETCD connection information within it, and running **kubectl create** against it.
5. ????
6. PROFIT!!!

Beyond the (ongoing) changes to this repo, I've created two new components, [layer-calico-cni](https://github.com/wwwtyro/layer-calico-cni) and [interface-kubernetes-cni](https://github.com/wwwtyro/interface-kubernetes-cni).

Here's how kubernetes-master, kubernetes-worker, calico-cni, and interface-kubernetes-cni work together:

#### kubernetes-master and calico-cni subordinate
- calico-cni installs calicoctl utility binary
- calico-cni relates to the master over the k8s-cni interface
- calico-cni relates to etcd
- calico-cni creates calicoctl systemd service with information from etcd relation
- calico-cni starts calicoctl systemd service
- master tells calico-cni it's a master subordinate

#### kubernetes-worker and calico-cni subordinate
- calico-cni installs calicoctl utility binary
- calico-cni relates to the worker over the k8s-cni interface
- calico-cni relates to etcd
- calico-cni creates calicoctl systemd service with information from etcd relation
- calico-cni starts calicoctl systemd service
- worker tells calico-cni it's a worker subordinate
- worker tells calico-cni the location of the kubeconfig file
- calico-cni installs **calico** and **calico-ipam** plugins
- calico-cni creates `/etc/cni/net.d/10-calico.conf` file using etcd connection information and kubeconfig location
- calico-cni renders the network policy controller manifest with etcd connection information
- calico-cni calls **kubectl apply** to instantiate the network policy controller

For reasons ranging from my slow wit to my impatience, the above differs from the current implementation in a few ways that should probably be addressed:

* The CNI lo plugin (**loopback**) is installed on the master subordinate. It probably doesn't need to be.
* The Calico CNI plugins (**calico** and **calico-ipam**) are installed on the master subordinate. They probably shouldn't be.
* I'm actually calling **kubectl create** to make the policy controller; it should be **apply** because idempotentcy.
* I'm pretty sure interface-kubernetes-cni is very wrong with respect to the conversation between units. I was saving fixing it for later, but now you're saddled with it. Sorry! :disappointed: 

A few notes:
* DNS *everything* has been removed (commented out mostly, in this PR) from the master and worker code. This was done because DNS was previously set up with information from the sdn interface. Needless to say, it will need to be reintroduced.
* You can find the bundle I've been testing with [here](https://gist.github.com/wwwtyro/b23f38c0f4fe92183cc3ae1daaf28989).
* The **calicoctl** utility kicks off a docker container, so layer-docker has been included. This means that if you're in a limited egress environment, you'll need to add the http(s)-proxy configuration. This is done in the bundle above, for example.
* Current testing is at the point of **calicoctl** running (though I see two processes - not sure why yet!) on the master and worker nodes. Port 6443 is exposed on the master, but it appears unreachable for some reason - figuring that out was my next step. Network policy controller deployment and CNI, you know, actually working, is still completely untested, but should be near at hand.

I *think* that's all the information I have to upload to you guys. If I remember anything else, I'll put it in a new comment here. @Cynerva has thoroughly absorbed (like the freakish sponge he is) all of this and is ready to take over. I'm extremely available next week; don't hesitate to reach out if you need some clarification.

Good luck, ryebot out.

### CODE REVIEW

**Comment by wwwtyro** [comment]
Forgot to mention: the docs recommend - but do not require - setting up an etcd cluster separate from the k8s cluster for Calico. After some discussion, that was not done here for reasons including expedience and convergence. The etcd cluster is currently shared between k8s and Calico.

**Comment by Cynerva** [comment]
This is in a working state, but with a lot of cleanup left to do. Still working on it.

@marcoceppi I believe you asked me yesterday how reusable the CNI portion of this is - or something to that effect. Here's an attempt to answer that.

Most communication occurs through CNI:

* calico charm installs `calico` and `calico-ipam` CNI plugins
* calico charm provides `10-calico.conf` CNI plugin configuration file
* kubelet is configured to use CNI, without needing to know anything about calico

Some communication occurs outside the scope of CNI:
* The `10-calico.conf` file provided by calico has [some kubernetes-specific configuration](https://github.com/wwwtyro/layer-calico-cni/blob/master/templates/10-calico.conf#L16)
* In order to support Kubernetes NetworkPolicy stuff, the calico charm must deploy [calico-policy-controller](https://github.com/wwwtyro/layer-calico-cni/blob/master/reactive/calico_cni.py#L146) into the cluster.

It is not clear to me if other CNI implementations will require similar, kubernetes-specific communication or not. However, based on the [NetworkPolicy prerequisites](http://kubernetes.io/docs/user-guide/networkpolicies/#prerequisites) from Kubernetes docs, I suspect every implementation will at least need to deploy its own policy controller.

**Comment by marcoceppi** [comment]
So, this looks really good. I think for `service_cluster_cidr` we should consider using extra-bindings/network-spaces to have operators attach spaces directly. Nothing for this inital PR, configuration option suffices for the time being.

**Comment by Cynerva** [comment]
Okay, I've done about as much cleanup here as I can without further feedback. This is ready for more thorough review.

Please review [interface-kubernetes-cni](https://github.com/wwwtyro/interface-kubernetes-cni) and [layer-calico](https://github.com/wwwtyro/layer-calico-cni) (repo name needs update) along with the changes here. Cheers!

File: cluster/juju/layers/kubernetes-master/config.yaml [hunk 1]
```diff
@@ -7,3 +7,7 @@ options:
     type: string
     default: cluster.local
     description: The local domain for cluster dns
+  service-cidr:
+    type: string
+    default: 10.152.183.0/24
+    description: CIDR to user for Kubernetes services. Cannot be changed after deployment.
```
Comment (Cynerva) [inline cluster/juju/layers/kubernetes-master/config.yaml]: Acknowledging the immutable config here. Changing the IP range in a live Kubernetes cluster doesn't look to be simple.

**Comment by Cynerva** [comment]
Test bundle uploaded as `cs:~cynerva/bundle/kubernetes-calico-0` ([bundle.yaml](https://api.jujucharms.com/charmstore/v5/~cynerva/bundle/kubernetes-calico-0/archive/bundle.yaml))

**Comment by Cynerva** [comment]
e2e results against kubernetes-calico, MAAS cloud, egress limited:

```
Summarizing 6 Failures:

[Fail] [k8s.io] ClusterDns [Feature:Example] [It] should create pod that uses dns [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2183

[Fail] [k8s.io] Port forwarding [k8s.io] With a server that expects a client request [It] should support a client that connects, sends data, and disconnects [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/portforward.go:154

[Fail] [k8s.io] Port forwarding [k8s.io] With a server that expects no client request [It] should support a client that connects, sends no data, and disconnects [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/portforward.go:154

[Fail] [k8s.io] Kubectl client [k8s.io] Kubectl run --rm job [It] should create a job from an image, then delete the job [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:1201

[Fail] [k8s.io] Port forwarding [k8s.io] With a server that expects a client request [It] should support a client that connects, sends no data, and disconnects [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/portforward.go:154

[Fail] [k8s.io] Networking [It] should provide Internet connection for containers [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/networking.go:54

Ran 99 of 385 Specs in 531.033 seconds
FAIL! -- 93 Passed | 6 Failed | 0 Pending | 286 Skipped 

Ginkgo ran 1 suite in 8m51.857514271s
Test Suite Failed
```

**Comment by Cynerva** [comment]
These scripts capture the process we used for building the calico resource, and updating the kubernetes-worker resource.

* [build-calico-resource.sh](https://gist.github.com/Cynerva/f24b02ef5a112d1251c95cff77e8fb44)
* [add-worker-loopback.sh](https://gist.github.com/Cynerva/d52c69dbdf018748948cecb1c29ace15)

No other resources were harmed in the making of this prototype.

File: cluster/juju/layers/kubernetes-master/config.yaml (hunk 1 above)
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/config.yaml]: Acknowledging immutable config sadly does not make it go away. And we all know immutable config breaks user experience. If this value can not be changed... Can we consider it a layer option?

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 2]
```diff
@@ -80,22 +85,30 @@ def install():
...
+        install = ['install', '-v', '-D', unpacked, app_path]
+        hookenv.log(install)
+        check_call(install)
 
     set_state('kubernetes-master.components.installed')
 
 
+@when('cni.connected')
+@when_not('cni.configured')
+def configure_cni(cni):
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: No docstring or comment here to tell me what is going on or why this code exists.

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 3]
```diff
@@ -219,29 +231,21 @@ def push_api_data(kube_api):
...
     ''' Launch dashboard addons if they are enabled in config '''
     if hookenv.config('enable-dashboard-addons'):
         hookenv.log('Launching kubernetes dashboard.')
         context = {}
         context['arch'] = arch()
-        context['pillar'] = {'num_nodes': get_node_count()}
-        for template in dashboard_templates:
-            create_addon(template, context)
-        set_state('kubernetes.dashboard.available')
+        try:
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Good idea on the try/except here!

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 4]
```diff
@@ -278,8 +281,15 @@ def start_kube_dns(sdn_plugin):
     hookenv.log(message)
     hookenv.status_set('maintenance', message)
 
-    context = prepare_sdn_context(sdn_plugin)
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Goodbye hacky old method. I much prefer the new way you build context. Great job here!

File: cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py [hunk 5]
```diff
@@ -133,29 +132,39 @@ def update_kubelet_status():
...
     dns = kube_dns.details()
+
     if (data_changed('kube-api-servers', servers) or
             data_changed('kube-dns', dns)):
         # Initialize a FlagManager object to add flags to unit data.
         opts = FlagManager('kubelet')
         # Append the DNS flags + data to the FlagManager object.
-        opts.add('--cluster-dns', dns['sdn-ip'])
+
+        opts.add('--cluster-dns', dns['sdn-ip']) # FIXME: sdn-ip needs a rename
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py]: Is the original code using the wrong value here? Or does the name just need CNI? I am all for renaming it as you see fit.

File: cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py [hunk 6]
```diff
@@ -284,11 +293,12 @@ def render_init_scripts(api_servers):
     kubelet_opts = FlagManager('kubelet')
     # Declare to kubelet it needs to read from kubeconfig
     kubelet_opts.add('--require-kubeconfig', None)
-    kubelet_opts.add('--kubeconfig', '/srv/kubernetes/config')
+    kubelet_opts.add('--kubeconfig', kubeconfig_path)
+    kubelet_opts.add('--network-plugin', 'cni')
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py]: This new flag is incompatible with software defined networking (which is totally fine) but we could put this line inside an if block when using CNI. I know that is not in the scope of this PR. Given your expertise: Would we be able to support both SDN and CNI with these charms?

File: cluster/juju/layers/kubernetes-master/config.yaml (hunk 1 above)
Comment (marcoceppi) [inline cluster/juju/layers/kubernetes-master/config.yaml]: I disagree about layer config, that means, you basically have to compile the charm to change the network CIDR if it conflicts with the network. I'd rather have the hook error saying "You can't change CIDR after deployment"

Comment (Cynerva) [inline cluster/juju/layers/kubernetes-master/config.yaml]: We'll see if we can have it output an appropriate error message. I think that's the best we can do for now.

File: cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py (hunk 6 above)
Comment (Cynerva) [inline cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py]: I think it'd be doable, but not trivial. If SDN and CNI are different interfaces, then we'd have to consider some weird cases - for example, what happens if we end up with relations to both?

For what it's worth, Flannel has a CNI plugin so my preference would be to drop SDN entirely and see if we can get Flannel working through CNI instead.

Comment (chuckbutler) [inline cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py]: We removed interface:sdn-plugin.  Why would we need to support both?

**Comment by Cynerva** [comment]
Pushed changes:
* Merged in master-node-split since we're a bit behind
* Added missing docstrings
* Added a warning message when service-cidr config changes

**Comment by Cynerva** [comment]
Created an awful script that captures the steps for deploying on AWS: [deploy-kubernetes-calico-aws.sh](https://gist.github.com/Cynerva/76999bd7c3e9571f2d8518127955b159)

File: cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py (hunk 6 above)
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py]: no need to support both, dropping it like its hot.

**Pull request merged**